### ISSUE TITLE
deleting a node does not update protoflux groups propertly

### ISSUE THREAD

**Issue by art0007i** [Yellow-Dog-Man/Resonite-Issues#907]
### Describe the bug?

When using ProtoFlux I noticed that sometimes selecting a node using the flux tip selects nodes from a completely disconnected group of nodes. I found out that this happens specifically when you have 1 group of nodes, and delete all the nodes that are connecting a specific part, the game will still consider these connected even though there is no actual connection.

### To Reproduce

1. Spawn any chain of nodes (in my case I used this resrec:///U-art0007i/R-d6278c17-5ee5-4ce4-8cf5-113b3ec470cb)
2. delete any node in the middle of the chain
3. select a node on either of the 2 newly created groups
4. both groups will be selected despite having nothing connecting them

### Expected behavior

selecting nodes only nodes which are connected to each other

### Screenshots

![nodebug](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/19620451/817e797a-f70f-4fec-bdca-2cda468bc144)

### Resonite Version Number

2023.12.15.22

### What Platforms does this occur on?

Windows

### Log Files

[ROTHORN - 2023.12.15.22 - 2023-12-15 16_09_19.log](https://github.com/Yellow-Dog-Man/Resonite-Issues/files/13686767/ROTHORN.-.2023.12.15.22.-.2023-12-15.16_09_19.log)

### Additional Context

it seems that when cutting a wire in either of the bugged groups fixes this issue

**Comment by art0007i** [comment]
All the nodes I've tried experience this problem (including ValueAdd)
It seems to happen to me without fail every single time. I really don't know why you can't replicate it...

Here's a video of me doing this bug from launch to exit in multiple different configurations:
https://youtu.be/1TJKyPWSpSc
[ROTHORN - 2023.12.15.22 - 2023-12-16 01_43_01.log](https://github.com/Yellow-Dog-Man/Resonite-Issues/files/13690889/ROTHORN.-.2023.12.15.22.-.2023-12-16.01_43_01.log)

**Comment by Frooxius** [comment]
This is an issue that came up before, but replicating this might be tricky.

This likely happens because of how the node grouping and group separation works, which is very sensitive to the specific order this happens in.

Simply changing the order in which the nodes are added, connected and unplugged can make this not happen, so replication will likely require the steps to be exact.

**Comment by Banane9** [comment]
I think the missing step in the replication is that you must select everything once before deleting the node and re-selecting to experience the problem. At least that's what you've done in every clip but isn't mentioned in the reproduction steps.

Also: nice exit screen 👌

**Comment by art0007i** [comment]
> I think the missing step in the replication is that you must select everything once before deleting the node and re-selecting to experience the problem. At least that's what you've done in every clip but isn't mentioned in the reproduction steps.


I don't know why I keep doing that thing, it's some kind of natural instinct but I have done it exactly once in the video at 2:55

**Comment by Banane9** [comment]
> I don't know why I keep doing that thing, it's some kind of natural instinct but I have done it without re-selecting once in the video at 2:55

Oh yeah, nevermind then.

> This likely happens because of how the node grouping and group separation works, which is very sensitive to the specific order this happens in.

Would it make sense then to dynamically traverse connected nodes when selecting to determine the group? 🤔

**Comment by Frooxius** [comment]
> Would it make sense then to dynamically traverse connected nodes when selecting to determine the group? 🤔

No, that would just mask the bug and we don't do that - we want to fix the root issue.

Those nodes are supposed to be separate groups internally once you split them up - meaning the algorithm that traverses them is supposed to split them up, but it's not doing that - we need to fix that.

**Comment by Nytra** [comment]
I have found a bit of a simpler test for this. Create a ProtoFlux setup of three If nodes like in this image:

![2024-02-24 13 34 19](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/14206961/4bcb8b77-fe7d-494c-8b78-27d9f334327d)

Cutting one of the two wires does not split the group, and all nodes will still be in the same group:

![2024-02-24 13 36 09](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/14206961/50ca63ad-ae07-42ab-94e4-c47648665f9c)

![2024-02-24 13 36 17](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/14206961/8ec5a46a-f939-4de2-a040-0c5ebc776a40)

In addition, grabbing and destroying the node on the left will not split the group, and the two nodes on the right will remain connected:

![2024-02-24 13 37 44](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/14206961/fcfe6fcc-0e53-4584-8714-3fb15601a789)

Cutting both wires does split the group:

![2024-02-24 13 38 56](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/14206961/2297b18f-1e93-465e-8e77-ae177c8c496f)

**Comment by Redd56** [comment]
can confirm it still occurs, just lost 30 minutes of work from this
